### PR TITLE
Add NotificationContext and useNotification hook

### PR DESCRIPTION
**Pull request by Griffin-Sullivan**
## Description
This PR adds a new global context for Notifications. I also added a `useNotification` hook to allow for easier dispatching by calling the specific method you want to use `notification.success` or `notification.error`. I also looped in the removal of notifications into this hook. You can call it like: 
```javascript
notifications = useNotification() // returns state for notifications
...
notifications.remove(notification.id)
```

image

* I still not have tested restoring an archived version yet since I can't do it in mocked mode

There are a couple known issues at the moment:
1. The id count gets reset to 0 whenever you change pages meaning if you switch pages quickly and get another notification they will have the same id
2. The notification exit button looks out of place (I think it should be center aligned not in the top right corner)

## Merge criteria:

- All the commits have been [_signed-off_](https://github.com/kubeflow/community/tree/master/dco-signoff-hook#signing-off-commits)  (To pass the `DCO` check)

- [x] The commits have meaningful messages; the author will squash them after approval or in case of manual merges will ask to merge with squash.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work.
- [x] Code changes follow the [kubeflow contribution guidelines](https://www.kubeflow.org/docs/about/contributing/).

If you have UI changes

- [x] The developer has added tests or explained why testing cannot be added.
- [x] Included any necessary screenshots or gifs if it was a UI change.
- [x] Verify that UI/UX changes conform the UX guidelines for Kubeflow.

### CODE REVIEW

File: clients/ui/frontend/src/app/hooks/useNotification.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+type NotificationRemoveProps = (id: number | undefined) => void;
+
+type NotificationTypeFunc = {
+  [key in NotificationTypes]: NotificationProps
+}
+
+interface NotificationFunc extends NotificationTypeFunc {
+  remove: NotificationRemoveProps
+}
```
Comment (Griffin-Sullivan) [inline clients/ui/frontend/src/app/hooks/useNotification.ts:23]: This is a little hacky. Maybe we should just have a separate hook for when we want to return all notifications and do something with them like removing one.

File: clients/ui/frontend/src/components/ToastNotification.tsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    <Alert
+      variant={asEnumMember(notification.status, AlertVariant) ?? undefined}
+      title={notification.title}
+      actionClose={
+        <AlertActionCloseButton onClose={() => notifications.remove(notification.id)} />
+      }
+      onMouseEnter={() => setMouseOver(true)}
+      onMouseLeave={() => setMouseOver(false)}
+    >
+      {notification.message}
```
Comment (Griffin-Sullivan) [inline clients/ui/frontend/src/components/ToastNotification.tsx:47]: See the description: this is not being passed along properly.

File: clients/ui/frontend/src/components/ToastNotifications.tsx [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import React, { useContext } from 'react';
+import { AlertGroup } from '@patternfly/react-core';
+import ToastNotification from './ToastNotification';
+import { NotificationContext } from '~/app/context/NotificationContext';
+
+const ToastNotifications: React.FC = () => {
+  const [notifications] = useContext(NotificationContext);
```
Comment (Griffin-Sullivan) [inline clients/ui/frontend/src/components/ToastNotifications.tsx:7]: The current hook only gives access to notification actions. Should we expand it to also return the notifications list? This is why I'm using `useContext` here.

File: clients/ui/frontend/src/utilities/utils.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+export const asEnumMember = <T extends object>(
```
Comment (lucferbux) [inline clients/ui/frontend/src/utilities/utils.ts:1]: I think this already existed in the codebase. We can either keep one or if you feel this should be in the main utils I'm up for it.

File: clients/ui/frontend/src/app/context/NotificationContext.tsx [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  title: string;
+  message?: React.ReactNode;
+  hidden?: boolean;
+  read?: boolean;
+  timestamp: Date;
+};
+
+type NotificationAction =
+  | {
+      type: 'add_notification';
```
Comment (lucferbux) [inline clients/ui/frontend/src/app/context/NotificationContext.tsx:16]: I would create a new enum to detail the types, it's a nice way to keep everything altogehter a part from the type itself.

File: clients/ui/frontend/src/app/context/NotificationContext.tsx [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  action,
+) => {
+  switch (action.type) {
+    case 'add_notification': {
+      return [
+        ...notifications,
+        {
+          status: action.payload.status,
+          title: action.payload.title,
+          timestamp: action.payload.timestamp,
```
Comment (lucferbux) [inline clients/ui/frontend/src/app/context/NotificationContext.tsx:46]: You are not passing the message here, that's why the message is not displayed.

File: clients/ui/frontend/src/components/ToastNotification.tsx (hunk 2 above)
Comment (lucferbux) [reply clients/ui/frontend/src/components/ToastNotification.tsx:45]: Left a comment with the fix

**Review by lucferbux — CHANGES_REQUESTED** [review]
Haven't tested it yet, but did a quick review, we can take a deeper look later today if you want.

**Comment by Griffin-Sullivan** [comment]
Ok @lucferbux addressed your comments thanks!

**Comment by Griffin-Sullivan** [comment]
Ok with the newest push I have added a bunch of testing which includes some for checking the notification pops up correctly.

File: clients/ui/frontend/src/__tests__/cypress/cypress/pages/components/Notification.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+export class Notification {
```
Comment (Griffin-Sullivan) [inline clients/ui/frontend/src/__tests__/cypress/cypress/pages/components/Notification.ts:1]: Maybe we should rename `ToastNotification`?

File: clients/ui/frontend/src/__tests__/cypress/cypress/tests/mocked/modelVersionArchive.cy.ts [hunk 8]
```diff
@@ -185,119 +188,124 @@ describe('Model version archive list', () => {
...
+    cy.interceptApi(
+      'PATCH /api/:apiVersion/model_registry/:modelRegistryName/model_versions/:modelVersionId',
+      {
+        path: {
+          modelRegistryName: 'modelregistry-sample',
+          apiVersion: MODEL_REGISTRY_API_VERSION,
+          modelVersionId: 2,
+        },
+      },
+      mockBFFResponse(mockModelVersion({})),
```
Comment (Griffin-Sullivan) [inline clients/ui/frontend/src/__tests__/cypress/cypress/tests/mocked/modelVersionArchive.cy.ts:202]: Notice how we have to put a bunch of `mockBFFResponse` everywhere. We should figure out a way for all of our `mockObject` calls to inherently know about the BFF response so we don't have to be so explicit.

File: clients/ui/frontend/src/__tests__/cypress/cypress/pages/components/Notification.ts (hunk 7 above)
Comment (lucferbux) [reply clients/ui/frontend/src/__tests__/cypress/cypress/pages/components/Notification.ts:1]: Yes, we are not using Notification, we might even refactor this in the future if this is not part of the project but reusable code, we'll see

Comment (lucferbux) [reply clients/ui/frontend/src/__tests__/cypress/cypress/pages/components/Notification.ts:1]: At the moment, just rename it and we can call it a day

File: clients/ui/frontend/src/__tests__/cypress/cypress/tests/mocked/modelVersionArchive.cy.ts (hunk 8 above)
Comment (lucferbux) [reply clients/ui/frontend/src/__tests__/cypress/cypress/tests/mocked/modelVersionArchive.cy.ts:202]: This is a great point, my first guess is to just modifify the mock objects and add a wrapper there, can you open a ticket in enhancemets to cover this?

File: clients/ui/frontend/src/app/context/NotificationContext.tsx [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+type NotificationAction =
+  | {
+      type: NotificationActionTypes.ADD_NOTIFICATION;
+      payload: Notification;
+    }
+  | {
+      type: NotificationActionTypes.DELETE_NOTIFICATION;
+      payload: { id: Notification['id'] };
+    };
```
Comment (lucferbux) [inline clients/ui/frontend/src/app/context/NotificationContext.tsx:27]: Can you move this to the `types.ts` file under `app`?

File: clients/ui/frontend/src/app/hooks/useNotification.ts [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+};
+
+interface NotificationFunc extends NotificationTypeFunc {
+  remove: NotificationRemoveProps;
+}
+
+export const useNotification = (): NotificationFunc => {
+  const { dispatch } = useContext(NotificationContext);
+  // need to move this count somewhere else since it will reset on every new useNotification instance (like switching pages)
+  let notificationCount = 0;
```
Comment (lucferbux) [inline clients/ui/frontend/src/app/hooks/useNotification.ts:27]: Why don't you move the notification count to the context itself?

**Review by lucferbux — APPROVED** [review]
Everything looks great, just a couple of nits, really great work here!

File: clients/ui/frontend/src/app/hooks/useNotification.ts (hunk 10 above)
Comment (Griffin-Sullivan) [reply clients/ui/frontend/src/app/hooks/useNotification.ts:27]: Why don't I......

**Review by ederign — APPROVED** [review]

**Pull request merged**